### ISSUE TITLE
BUG: DropoutWrapper incorrectly updates memory state

### ISSUE THREAD

**Issue by santon834** [tensorflow/tensorflow#11650]
### Describe the problem
GRU units (wikipedia):
![image](https://user-images.githubusercontent.com/26974149/28431669-fb89275a-6d39-11e7-929a-43be917d2602.png)

tf.contrib.rnn.DropoutWrapper documentation:
state_keep_prob: unit Tensor or float between 0 and 1, output keep probability; if it is constant and 1, no output dropout will be added. State dropout is performed on the output states of the cell.

In GRU units, the output state is the memory state. When variational_recurrent=True, the same temporal dropout mask is applied to the output state, in each time step, with the remaining outputs divided by the dropout probability. This leads to exponential growth of the memory state and exploding outputs (given long enough time series):
![image](https://user-images.githubusercontent.com/26974149/28432360-7644d7b2-6d3c-11e7-8999-80ff55344f61.png)

The correct way is probably to divide U_z, U_r and U_h and not the output state.

### Source code / logs
```
################################### 
def length(sequence):
	used = tf.sign(tf.reduce_max(tf.abs(sequence), axis=2))
	length = tf.reduce_sum(used, axis=1)
	return length
################################### 
def GRU(x, units, act, in_dp, out_dp, tmp_dp):
	gru_cell = tf.contrib.rnn.GRUCell(units, activation=act, kernel_initializer=tf.contrib.layers.xavier_initializer(uniform=True))
	wrapped_gru_cell = tf.contrib.rnn.DropoutWrapper(gru_cell, input_keep_prob=in_dp, output_keep_prob=out_dp, state_keep_prob=tmp_dp, variational_recurrent=True, dtype=x.dtype, input_size=x.get_shape()[2])
	outputs, state = tf.nn.dynamic_rnn(wrapped_gru_cell, x, dtype=x.dtype, sequence_length=length(x))
	return [outputs, state]
################################### 
Running with parameters (all exploded):
act=tf.nn.tanh, in_dp=1., out_dp=1., tmp_dp=0.5/0.9, n_hidden = 32/64/128
```

### System information
== cat /etc/issue ===============================================
Linux mvdslab 4.4.0-83-generic #106-Ubuntu SMP Mon Jun 26 17:54:43 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux
VERSION="16.04.2 LTS (Xenial Xerus)"
VERSION_ID="16.04"
VERSION_CODENAME=xenial

== are we in docker =============================================
No

== compiler =====================================================
c++ (Ubuntu 5.4.0-6ubuntu1~16.04.4) 5.4.0 20160609
Copyright (C) 2015 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

== uname -a =====================================================
Linux mvdslab 4.4.0-83-generic #106-Ubuntu SMP Mon Jun 26 17:54:43 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux

== check pips ===================================================
numpy (1.12.1)
numpydoc (0.6.0)
protobuf (3.3.0)
tensorflow (1.2.0)
tensorflow-gpu (1.2.0)

== check for virtualenv =========================================
False

== tensorflow import ============================================
tf.VERSION = 1.2.0
tf.GIT_VERSION = v1.2.0-1131-gbc691dd
tf.COMPILER_VERSION = v1.2.0-1131-gbc691dd
Sanity check: array([1], dtype=int32)

== env ==========================================================
LD_LIBRARY_PATH /home/anton/torch/install/lib:/usr/local/cuda-8.0/lib64
DYLD_LIBRARY_PATH /home/anton/torch/install/lib:/home/anton/torch/install/lib:

== nvidia-smi ===================================================
Thu Jul 20 10:36:06 2017       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 375.66                 Driver Version: 375.66                    |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  Quadro K1200        Off  | 0000:01:00.0      On |                  N/A |
| 39%   38C    P0     1W /  35W |    459MiB /  4034MiB |      3%      Default |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID  Type  Process name                               Usage      |
|=============================================================================|
|    0      1183    G   /usr/lib/xorg/Xorg                             110MiB |
|    0      6218    G   ...el-token=544D2517E333A364F4E0C0630D2C9DD1   347MiB |
+-----------------------------------------------------------------------------+

== cuda libs  ===================================================
/usr/local/cuda-8.0/lib64/libcudart.so.8.0.61
/usr/local/cuda-8.0/lib64/libcudart_static.a
/usr/local/cuda-8.0/doc/man/man7/libcudart.7
/usr/local/cuda-8.0/doc/man/man7/libcudart.so.7

**Comment by santon834** [comment]
Update:
The same problem occurs with LSTM:
![image](https://user-images.githubusercontent.com/26974149/28434593-b7d913ca-6d44-11e7-9429-c7d6bc75ed81.png)

c_t-1 is divided by the dropout probability and grows exponentially instead of dividing the matrices U

**Comment by santon834** [comment]
Update:
The same problem occurs with variational_recurrent=false.
Line 754 in rnn_cell_impl.py updates the memory state with the dropout mask and divides the state by the probability, which is wrong. The division in dropout is performed to adjust for a constant expected activation arguments. However, it should not modify the internal memory state of the unit. 
With variational_recurrent=false, we have a random recurrent dropout mask at each step and we don't get the same "accumulating effect" as with variational_recurrent=true. As a result, the memory state remains in the same magnitude as the candidate update (right hand element in the memory state update equation) and is practically impossible to debug. I wonder if this is the cause for the general consensus that recurrent dropout doesn't work/won't remember.

**Comment by asimshankar** [comment]
@ebrevdo : Mind taking a look?

**Comment by ebrevdo** [comment]
Will need to confirm with folks internally, but we may have to add an exclusions field to the wrapper to avoid dropping out the c state.

**Comment by davidhstern** [comment]
Has anyone had a chance to look at this?  I agree with @santon834 that it looks like variational dropout  has been wrongly implemented for the LSTM and GRU if state_keep_prob < 1.  I think that the LSTM could be fixed by excluding the c state from dropout.   However, fixing the GRU would require more changes as dropout should only be applied after the weights (U) are applied but not after the multiplication by z.   I am not sure the best way to achieve this but perhaps a new pair of cells with dropout built in might be the way forward?

I also agree that it the implemetation seems to give poor results even when variational_recurrent=False and it might work better if it did not drop the internal LSTM state.   However, I guess the best implementation of non-variational dropout for LSTM is more open to debate so I would not call that part wrong necessary.

**Comment by davidhstern** [comment]
@santon834 You might want to check out https://arxiv.org/pdf/1603.05118.pdf

It looks like not dropping the LSTM memory state might actually be the key to getting good performance from dropout in LSTMs.  Using one mask per sequence might be redundant as long as you do this.

**Comment by ebrevdo** [comment]
I'm testing a fix for LSTM-type cells now.  GRU, as you suggested, is
harder to fix.

**Comment by davidhstern** [comment]
FWIW, I found that simply applying regular dropout to the c tensor in the GRUCell gives good results, as suggested by the paper linked above.   Obviously this meant modifying the GRUCell.  I am not sure how to do it in the wrapper.

**Comment by ebrevdo** [comment]
That's impossible to do in the wrapper; I think one would have to add a new
GRUCell type to get this behavior.

**Comment by dchatterjee172** [comment]
@ebrevdo Is this implemented for GRUs?
